rspamd: add feature for adjusting options with a file parsed by DMS

georglauterbach opened the pull request:
# Description

Users should be able to do some easy adjustments with our help. With this addition, they can mount a file `rspamd-modules.conf` and our setup will evaluate it. You can read the documentation for this when looking at the docs preview for this PR :)

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (README.md or the documentation under `docs/`)
- [x] If necessary I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

polarathene commented on target/scripts/startup/setup-stack.sh:
```diff
@@ -141,6 +141,64 @@ EOF
+
+    while read -r COMMAND MODULE ARGUMENT1 ARGUMENT2
+    do
+      case "${COMMAND}"
+      in
```
```suggestion
      case "${COMMAND}" in
```

polarathene reviewed (changes requested):
Looks fine :)

The docs could probably do with some revision, but I lack the time to go through that. They're good enough.

I'm not sure about `rspamd-commands` as the config file name. Is it just for managing modules? Why not something like `rspamd.mod`, `rspamd-modules.conf` (_potentially gets syntax highlighting for comments in an editor_), or something else that looks more like a text file rather than a binary or folder at a glance? :sweat_smile: 

Likewise, `config/override-configs` was for a specific test to use. With the intent of having a cleaner separation of what files belong to what tests. I haven't given that too much thought, so it might make sense for some tests to share other config files that aren't in the common default config in tests.

georglauterbach commented on test/helper/sending.bash:
```diff
@@ -56,11 +56,12 @@ function _send_mail_and_get_id() {
   MAIL_ID=$(_exec_in_container tac /var/log/mail.log              \
     | grep -E -m 1 'postfix/smtpd.*: [A-Z0-9]+: client=localhost' \
-    | grep -E -o '[A-Z0-9]{11}')
+    | grep -E -o '[A-Z0-9]{9,12}' || true)
```
Forgot to mention this (FYI): On aarch64, the ID postfix uses might not be 11 characters long, so I provide a bit of range here.

georglauterbach replied:
I don't have the slightest idea.. but this is what I saw today on my arm64 machine.

polarathene commented on target/scripts/build/packages.sh:
```diff
@@ -132,16 +132,31 @@ function _install_dovecot
+  local DEB_FILE='/etc/apt/sources.list.d/rspamd.list'
+  local RSPAMD_PACKAGE_NAME
+
+  # We try getting the most recent version of Rspamd for aarch64 (from an official source, which
+  # is the backports repository). The version for aarch64 is 3.2; the most recent version for amd64
+  # that we get with the official PPA is 3.4.
+  #
+  # Not removing it later is fine as you have to explicitly opt into installing a backports package
+  # which is not something you could be doing by accident.
+  if [[ $(uname --machine) == 'aarch64' ]]
```
Not a big fan of this due to version difference, and the official rspamd advice is to not use the debian packages :sweat_smile: 

Hopefully the version disparity doesn't cause issues going forward if we take the approach you've chosen to support ARM64.

---

I had a browse of the [rspamd download docs](https://rspamd.com/downloads.html), and noticed they have support for Fedora (_and OpenSUSE via OBS, but TW isn't a good idea, and Leap releases are slow - not that Debian is any better_). 

At one point I wanted to consider Alpine as a base image offering, especially for the regular releases and newer packages, but I'm more reluctant now with all the `musl` compatibility and performance issues I've seen with it :sweat_smile: 

Fedora might be interesting, they have a minimal base image (slightly bigger than Debian), but hard to say if it's worthwhile or a smooth transition to change to. Those rspamd docs made it seem like it was supported but [it's been broken for a while](https://github.com/rspamd/rspamd.com/issues/604).

Maybe [OpenSUSE Leap image base](https://hub.docker.com/r/opensuse/leap/tags) would be worth looking into. It's fairly small at a glance, but I understand there is more familiarity with Debian :sweat_smile:  (_Leap is not likely to stay around either AFAIK as they're working towards a new container focused distro/spin_)

---

One other option is building rspamd ourselves with a multi-stage build, but that has it's own drawbacks (_maintenance + build time + CI cache?_).

georglauterbach replied:
I agree, it's not very nice, but as of now, ARM64 users would have to use v2.7. With this change, they can use v3.2 which is only 2 minor versions behind the official PPA. I will rework this as soon as a better emerges :)

---

About the base image change: we should put this into our backlog for now I think :)

polarathene replied:
> About the base image change: we should put this into our backlog for now I think :)

I am in no rush, that's not a fun change to tackle and really needs to be worthwhile to bother with :sweat_smile: 

My main motivations were attempting to reduce image weight and switch to something with more frequent release cadence for us to not rely on all these external builds for newer versions, as reliance on that seems to be growing.

---

Debian Bookworm release should bump up the rspamd package release at least? :)

I [raised a request to get official ARM64 builds supported](https://github.com/rspamd/rspamd/issues/4390), but given the history of maintainer stances on that, I'm not sure if it'll be approved on their end.

I've also further investigated the Fedora situation further and no fedora release going back 5 years has support to install `rspamd` based on their instructions :confused: Their docs are terribly out of date for that support it seems.

georglauterbach replied:
This is going to be a tough one indeed. Ubuntu probably wont help here either as it's basically Debian.. We should discuss about this first on a maintainers discussion board or with discussions for this repository so users can post too. 

I thought Fedora was actually quite nice, while I don't have a great experience with OpenSUSE, but I'm open to listening to all proposals.

polarathene replied:
No rush, we'll likely update to Bookworm before we change distro :)

polarathene reviewed (approved):
LGTM :)

georglauterbach commented:
Added a comment about the mail ID in e5ab05f2a55ebcc6e447711964aa2c7709f51d0d.

georglauterbach commented:
@polarathene I rebased on `master` and added more commits after caabadad18b7623eabe9422fc046fa8b54fbabf2. Just FYI when reviewing again. Since I updated the setup file quite significantly, its probably worthwhile to revisit the whole file for review again though. Configuration wise, not much has changes, but two more module configuration commands were added (and one command name was adjusted). More than that, there were proper additions to the tests & revision of the documentation.

georglauterbach commented on target/scripts/startup/setup-stack.sh:
```diff
@@ -702,9 +612,15 @@ function _setup_dkim_dmarc
     # shellcheck disable=SC2016
     sed -i -E 's|^(smtpd_milters =.*)|\1 \$dmarc_milter|g' /etc/postfix/main.cf
+  else
+    sed -i '/.*inet:localhost:8891/d' /etc/postfix/main.cf
```
Ah, forgot this in the commit message: our current `main.cf` still has the varibales `dkim_milter`/`dmarc_milter` and postconf will complain about them not being used when `ENABLE_OPENDKIM=0`/`ENABLE_OPENDMARC=0`. These changes (and line 619-623) get rid of the line and the warning.

polarathene replied:
Isn't this just for `dkim_milter` and `dmarc_filter`? Not better to match on those keys for deleting instead?

```suggestion
    sed -i '/^dkim_milter/d' /etc/postfix/main.cf
```

polarathene commented on docs/content/config/security/rspamd.md:
```diff
@@ -2,37 +2,87 @@
 ---
 
-!!! warning "Implementation of Rspamd into DMS is WIP!"
+!!! warning "The Current State of Rspamd's Integration Into DMS"
+
+    Recent pull requests have stabilized Rspamd's integration to a point where we encourage users to test the feature. We are confident that there are no major bugs in our code, that make using Rspamd infeasible, anymore. Please note that there may still be (breaking) changes ahead as integration is still work in progress!
```
```suggestion
!!! warning "The current state of Rspamd integration into DMS"

    Recent pull requests have stabilized integration of Rspamd to a point that we encourage users to test the feature. We are confident that there are no major bugs in our integration that make using Rspamd infeasible. Please note that there may still be (breaking) changes ahead as integration is still work in progress!
```

georglauterbach replied:
Done in 62e528d380ac92f6cc2e361c4685c01780d38088

polarathene commented on docs/content/config/security/rspamd.md:
```diff
@@ -2,37 +2,87 @@
 2. (re-)place files at `/etc/rspamd/local.d/` to override DMS settings and merge them with Rspamd settings
 
-You can find a list of all Rspamd modules [on their website][modules].
+inside the container.
```
Was this meant to be part of line 50?

georglauterbach replied:
Done in 62e528d380ac92f6cc2e361c4685c01780d38088

polarathene commented on target/scripts/helpers/setup-rspamd.sh:
```diff
@@ -0,0 +1,172 @@
+  then
+    __rspamd__log 'debug' 'Enabling ClamAV integration'
+    sedfile -i -E 's|^(enabled).*|\1 = true;|g' /etc/rspamd/local.d/antivirus.conf
+    # RSpamd uses ClamAV's UNIX socket, and to be able to read it, it must be in the same group
+    usermod -a -G clamav _rspamd
```
Should this not be handled the same way Amavis does in the Dockerfile? (_or the Amavis handling moved to it's setup script for enabling?_)

georglauterbach replied:
I'd rather do it here since we only really need it when using ClamAV. I'd advocate for moving the Amavis-handling into scripts then (follow-up PR).

polarathene commented on target/scripts/startup/setup-stack.sh:
```diff
@@ -640,9 +612,15 @@ function _setup_dkim_dmarc
   fi
 
-  [[ ${ENABLE_OPENDKIM} -eq 1 ]] || return 0
+  if [[ ${ENABLE_OPENDKIM} -eq 0 ]]
+  then
+    sed -i '/.*inet:localhost:8893/d' /etc/postfix/main.cf
+    return 0
+  fi
```
Same here, but note you're currently checking if OpenDKIM is disabled, while removing the `dmarc_milter` setting?

georglauterbach replied:
Done in cef100363fd6765954a7dd0af3ed76b3da8b4d01. Took a slightly different approach now: commented the milter and uncomment them when DKIM/DMARC is enabled.

georglauterbach commented:
Merged current `master` and resolved merge conflicts in [4a570ab](https://github.com/docker-mailserver/docker-mailserver/pull/3059/commits/4a570ab87321ec0ddc5eb525dd32d7d39c73879f). Added one more command (to adjust `options.inc`) in [96baf86](https://github.com/docker-mailserver/docker-mailserver/pull/3059/commits/96baf86acf41f36cd2bff0342d44364a86b4b3fe).

I will take care of resolving the conflicts that will arise when #3082 is merged. Then the PR is ready for review again :)

georglauterbach commented:
[1056876](https://github.com/docker-mailserver/docker-mailserver/pull/3059/commits/10568761479d6a84bc178ff20b73125adbe7428c) goes to show what I added to the docs a short while ago: "Program testing can be used to show the presence of bugs, but never to show their absence!" -- Dijkstra

polarathene reviewed (approved):
Praise be https://github.com/docker-mailserver/docker-mailserver/pull/3059/commits/96baf86acf41f36cd2bff0342d44364a86b4b3fe ! :rofl: 

> https://github.com/docker-mailserver/docker-mailserver/commit/10568761479d6a84bc178ff20b73125adbe7428c goes to show what I added to the docs a short while ago

Ha! I remember that quote from about a decade ago when I trained to be a software tester :sweat_smile: There was some terrifying bugs in medical equipment :grimacing:

georglauterbach commented:
> Praise be https://github.com/docker-mailserver/docker-mailserver/pull/3059/commits/96baf86acf41f36cd2bff0342d44364a86b4b3fe ! :rofl: 

😉

> > https://github.com/docker-mailserver/docker-mailserver/commit/10568761479d6a84bc178ff20b73125adbe7428c goes to show what I added to the docs a short while ago
>
> Ha! I remember that quote from about a decade ago when I trained to be a software tester :sweat_smile: There was some terrifying bugs in medical equipment :grimacing: 

Interesting! Never thought about hearing something linke this though.. actually a bit scary 🙈

georglauterbach commented:
Alright, after dismissing @polarathene's review for the one millionth time, I promise to not do it again (unless required by resolving another conflict - I do hope I don't need to..). I added proper commit messages to the last two commits to make the final review as easy as possible.

polarathene commented on target/scripts/startup/setup-stack.sh:
```diff
@@ -668,18 +645,27 @@ function _setup_dkim_dmarc
   then
     _log 'trace' "Adding OpenDMARC to Postfix's milters"
-
+    # Make sure to append the milter _behind_ the OpenDKIM milter!
```
Behind? After is probably more accurate? :)

```suggestion
    # Make sure to append the OpenDMARC milter _after_ the OpenDKIM milter!
```

georglauterbach replied:
Yeah, you're right.. :D

georglauterbach replied:
Done in [9756028](https://github.com/docker-mailserver/docker-mailserver/pull/3059/commits/97560282a7e640e80b9160df781366ec019744d0)

polarathene replied:
The comment suggested change? It didn't seem to get applied :thinking:

georglauterbach replied:
Oh my, sorry! I know why, a stupid mistake. Will push one more commit..

polarathene reviewed (dismissed):
> I decided against the proposed solution of using an array/string that is appended to Postfix's milters at the end due to unnecessarry complexity it would add IMO.

I'm guessing the complication was with expanding the array values within the sed command? Here's two different ways that it could work:

```bash
local DKIM_DMARC=()
# In conditionals:
DKIM_DMARC+='$dkim_milter'
DKIM_DMARC+='$dmarc_milter'

# After conditionals retrieve setting value and expand array into it
# (Note querying settings will probably delay without calling our utility method to avoid that):
postconf "smtpd_milters = $(postconf -h smtpd_milters) ${DKIM_DMARC[@]}"
# Alternatively, create a custom setting to use as an interpolated variable (by Postfix):
postconf "dms_dkim_dmarc = ${DKIM_DMARC[@]}"
sed -i -E 's|^(smtpd_milters =.*)|\1 \$dms_dkim_dmarc|g' /etc/postfix/main.cf

# If you ever need to retrieve the value
# Query the custom setting for value (-h) directly:
postfix -h dms_dkim_dmarc
# Or expand (-x) to the interpolated value where it's used:
postfix -hx smtpd_milters
```

I won't block on it, but it seemed a bit more DRY, in that the related logic between the two ENV conditions handled that shared setting together in one place afterwards :sweat_smile:

georglauterbach commented:
> > I decided against the proposed solution of using an array/string that is appended to Postfix's milters at the end due to unnecessarry complexity it would add IMO.
> 
> I'm guessing the complication was with expanding the array values within the sed command?

Actually, no :) I think your solution adds lines of code that are not necessary, since in your example you don't account for the extra calls to `_adjust_mtime_for_postfix_maincf` or the shellcheck comments. I implemented your solution, just to see what it looks like, but the extra complexity was not really worth it IMO. Moreover, there is no code handling the case in which `DKIM_DMARC` is empty, which adds another `if`..

> Here's two different ways that it could work: [...]

I do like some of the parts; especially adding things via `popstconf`. I have apllied some of this in the last PR feedback commit (97560282a7e640e80b9160df781366ec019744d0) :)

> I won't block on it, but it seemed a bit more DRY, in that the related logic between the two ENV conditions handled that shared setting together in one place afterwards 😅

I actually think it is really DRY now, and the current setup is easy to understand.

georglauterbach commented:
@polarathene this is a rather big PR, but I actually want to have this merged anyway. Do you think this PR is fine to merge with 1 review?

polarathene commented:
> Do you think this PR is fine to merge with 1 review?

I have been reasonably thorough with reviews. I know I sometimes miss things though. Pretty much our only other reviewer is Casper and he seems rather busy lately, probably swamped with notifications from this project to wade through too.

I don't see any harm in merging it, no release has been made, so there is still time to review even after a merge and provide feedback that can be addressed via a follow-up PR.

georglauterbach commented:
Alright, the I will go ahead :D

casperklein commented:
>  probably swamped with notifications from this project to wade through too.

Absolutely! lol
